compiler/catala_utils/messages.ml

```diff
@@ -2,188 +2,232 @@
+  let tags, str =
+    match target with
+    | Debug -> [Bold; Fg (C4 magenta)], "[DEBUG]"
+    | Error -> [Bold; Fg (C4 red)], "[ERROR]"
+    | Warning -> [Bold; Fg (C4 yellow)], "[WARNING]"
+    | Result -> [Bold; Fg (C4 green)], "[RESULT]"
+    | Log -> [Bold; Fg (C4 black)], "[LOG]"
+  in
+  if target = Debug then print_time_marker ppf ();
+  Format.pp_open_stag ppf (Ocolor_format.Ocolor_styles_tag tags);
+  Format.pp_print_string ppf str;
+  Format.pp_close_stag ppf ()
 
-let result_format format =
-  Format.printf ("%a" ^^ format ^^ "\n%!") result_marker ()
+(**{2 Printers}*)
 
-let log_format format =
-  Format.printf ("%a@[<hov>" ^^ format ^^ "@]@.") log_marker ()
+(** Prints the argument after the correct marker and to the correct channel *)
+let format target format =
+  let ppf = get_ppf target in
+  pp_marker target ppf ();
+  Format.pp_print_space ppf ();
+  Format.pp_open_hovbox ppf 0;
+  Format.kfprintf
+    (fun ppf ->
+      Format.pp_close_box ppf ();
+      Format.pp_print_newline ppf ())
+    ppf format
 
 (** {1 Message content} *)
 
 module Content = struct
-  type position = { message : string option; position : Pos.t }
-  type t = { message : string; positions : position list }
+  type message = Format.formatter -> unit
+  type position = { pos_message : message option; pos : Pos.t }
+  type t = { message : message; positions : position list }
+
+  let of_message (message : message) : t = { message; positions = [] }
 
-  let of_message (s : string) : t = { message = s; positions = [] }
+  let of_string (s : string) : t =
+    { message = (fun ppf -> Format.pp_print_string ppf s); positions = [] }
 end
 
+open Content
+
 let internal_error_prefix =
   "Internal Error, please report to \
    https://github.com/CatalaLang/catala/issues: "
 
 let to_internal_error (content : Content.t) : Content.t =
-  { content with message = internal_error_prefix ^ content.message }
-
-type content_type = Error | Warning | Debug | Log | Result
-
-let emit_content (content : Content.t) (typ : content_type) : unit =
-  let { Content.message = msg; positions = pos } = content in
+  {
+    content with
+    message =
+      (fun ppf ->
+        Format.fprintf ppf "%s@\n%t" internal_error_prefix content.message);
+  }
+
+let emit_content (content : Content.t) (target : content_type) : unit =
+  let { message; positions } = content in
   match !Cli.message_format_flag with
   | Cli.Human ->
-    (match typ with
-    | Warning -> warning_format
-    | Error -> error_format
-    | Debug -> debug_format
-    | Log -> log_format
-    | Result -> result_format)
-      "%s%s%s" msg
-      (if pos = [] then "" else "\n\n")
-      (String.concat "\n\n"
-         (List.map
-            (fun (pos : Content.position) ->
-              Printf.sprintf "%s%s"
-                (match pos.message with None -> "" | Some msg -> msg ^ "\n")
-                (Pos.retrieve_loc_text pos.position))
-            pos))
+    format target "@[<v>%t%a@]" message
```

Comment:
Wouahou what's `%t` ?

Reply:
Sorry if it's a bit obscure, it's a function that just takes a ppf as argument. In the above `"%t" message` is basically equivalent to `"%a" (fun ppf () -> message ppf) ()`

Reply:
Thanks, didn't know!